web/caret_browsing.js

```diff
@@ -0,0 +1,321 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Used to compare floats: there is no exact equality due to rounding errors.
+const PRECISION = 1e-1;
+
+class CaretBrowsingMode {
+  #mainContainer;
+
+  #toolBarHeight;
+
+  #viewerContainer;
+
+  constructor(mainContainer, viewerContainer, toolbarContainer) {
+    this.#mainContainer = mainContainer;
+    this.#viewerContainer = viewerContainer;
+    this.#toolBarHeight = toolbarContainer.getBoundingClientRect().height;
+  }
+
+  /**
+   * Return true if the two rectangles are on the same line.
+   * @param {DOMRect} rect1
+   * @param {DOMRect} rect2
+   * @returns {boolean}
+   */
+  #isOnSameLine(rect1, rect2) {
+    const top1 = rect1.y;
+    const bot1 = rect1.bottom;
+    const mid1 = rect1.y + rect1.height / 2;
+
+    const top2 = rect2.y;
+    const bot2 = rect2.bottom;
+    const mid2 = rect2.y + rect2.height / 2;
+
+    return (top1 <= mid2 && mid2 <= bot1) || (top2 <= mid1 && mid1 <= bot2);
+  }
+
+  /**
+   * Return true if the rectangle is under/over the caret.
+   * @param {DOMRect} rect
+   * @param {number} x
+   * @param {number} y
+   * @param {boolean} isUp
+   * @returns {boolean}
+   */
+  #isUnder(rect, x, y, isUp) {
+    const midY = rect.y + rect.height / 2;
+    return (
+      (isUp ? y >= midY : y <= midY) &&
+      rect.x - PRECISION <= x &&
+      x <= rect.right + PRECISION
+    );
+  }
+
+  /**
+   * Check if the rectangle is visible.
+   * @param {DOMRect} rect
+   * @returns {boolean}
+   */
+  #isVisible(rect) {
+    return (
+      rect.top >= this.#toolBarHeight &&
+      rect.left >= 0 &&
+      rect.bottom <=
+        (window.innerHeight || document.documentElement.clientHeight) &&
+      rect.right <= (window.innerWidth || document.documentElement.clientWidth)
+    );
+  }
+
+  /**
+   * Get the position of the caret.
+   * @param {Selection} selection
+   * @param {boolean} isUp
+   * @returns {Array<number>}
+   */
+  #getCaretPosition(selection, isUp) {
+    const { focusNode, focusOffset } = selection;
+    const range = document.createRange();
+    range.setStart(focusNode, focusOffset);
+    range.setEnd(focusNode, focusOffset);
+    const rect = range.getBoundingClientRect();
+
+    return [rect.x, isUp ? rect.top : rect.bottom];
+  }
+
+  static #caretPositionFromPoint(x, y) {
```

Comment:
Does this "need" to be a static method, or is there a reason for this that I'm missing?

Reply:
It doesn't need to have an instance and it's just an helper function. I wrote it for the Chrome case in order to make the code slightly more readable.